### PR TITLE
MSPB-71: Add office holidays import from CSV

### PR DESCRIPTION
**Pull request by pcandia**

### CODE REVIEW

File: submodules/strategyHolidays/strategyHolidays.js [hunk 1]
```diff
@@ -456,6 +576,56 @@ define(function(require) {
...
+		 * @param  {Array} csvData
+		 * @return {Array[]}
+		 * Intervals are extracted from CSV data and are not necessarily linear and/or concurrent.
+		 */
+
+		strategyHolidaysExtractDatesFromCsvData: function(csvData) {
+			var self = this,
+				months = _.map(self.i18n.active().strategy.holidays.months, function(value, key) {
+					return _.toLower(key);
+				});
```
Comment (azefiel) [inline submodules/strategyHolidays/strategyHolidays.js:590]: This mapping could be made more declarative, as we only want to operate on the keys:

```suggestion
				months = _
					.chain(self.i18n.active().strategy.holidays.months)
					.keys()
					.map(_.toLower)
					.value()
```

File: submodules/strategyHolidays/strategyHolidays.js [hunk 2]
```diff
@@ -456,6 +576,56 @@ define(function(require) {
...
+						id: 'new-' + Date.now(),
+						name: row.name,
+						recurring: row.recurring
+					},
+					endYear = row.year !== 0 ? row.year : new Date().getFullYear(),
+					dateArray = _.split(row.start_date, ' ');
+
+				if (!row.recurring) {
+					holidayData.endYear = endYear;
+				}
```
Comment (azefiel) [inline submodules/strategyHolidays/strategyHolidays.js:607]: Conditionally adding keys to an object can be done on variable assignment:

```js
var holidayData = _.merge({
	id: 'new-' + Date.now(),
	name: row.name,
	recurring: row.recurring
}, !row.recurring && {
	enYear: endYear
});
```

Using this pattern gives the advantage to expose possible props an object can have all in one place.

File: submodules/strategyHolidays/strategyHolidays.js [hunk 3]
```diff
@@ -456,6 +576,56 @@ define(function(require) {
...
+
+				holidayData.fromMonth = _.indexOf(months, dateArray[0]) + 1;
```
Comment (azefiel) [inline submodules/strategyHolidays/strategyHolidays.js:609]: Same here, the `fromMonth` prop should be declared on object assignment.

File: submodules/strategyHolidays/strategyHolidays.js [hunk 4]
```diff
@@ -456,6 +576,56 @@ define(function(require) {
...
+					holidayData.wday = dateArray[2];
+				} else {
+					holidayData.fromDay = _.toInteger(dateArray[1]);
+				}
+
+				if (row.type === 'range') {
+					var endDateArray = _.split(row.end_date, ' ');
+					holidayData.toMonth = _.indexOf(months, endDateArray[0]) + 1;
+					holidayData.toDay = _.toInteger(endDateArray[1]);
+				}
```
Comment (azefiel) [inline submodules/strategyHolidays/strategyHolidays.js:621]: Those conditional assignments are somewhat hard to follow, which does not makes it very straightforward to tell at a glance which props correspond to which holiday type. This complexity could be abstracted in a data structure:

```js
var configsPerType = {
	advanced: {
		ordinal: '',
		wday: ''
	},
	range: {
		fromDay: '',
		toMonth: '',
		toDay: ''
	},
	single: {
		fromDay: ''
	}
};
var typeConfig = _.get(configsPerType, row.type);
```
The result of which can then be added to our merge down of `holidayData`:
```js

var holidayData = _.merge({
	id: 'new-' + Date.now(),
	name: row.name,
	recurring: row.recurring
}, !row.recurring && {
	enYear: endYear
}, typeConfig);
```

File: submodules/strategyHolidays/strategyHolidays.js [hunk 5]
```diff
@@ -456,6 +576,56 @@ define(function(require) {
...
+				holidayRule.holidayData = holidayData;
```
Comment (azefiel) [inline submodules/strategyHolidays/strategyHolidays.js:622]: Now that the resolution of `holidayData` is just a merge down of objects, assignment could be moved during `holidayRule`'s declaration.

File: submodules/strategyHolidays/strategyHolidays.js [hunk 6]
```diff
@@ -456,6 +576,56 @@ define(function(require) {
 			return holidaysData;
 		},
 
+		/**
+		 * Returns an array of arrays containing intervals for each day of the week.
+		 * @param  {Array} csvData
+		 * @return {Array[]}
+		 * Intervals are extracted from CSV data and are not necessarily linear and/or concurrent.
```
Comment (azefiel) [inline submodules/strategyHolidays/strategyHolidays.js:583]: Not sure this JSDoc is relevant here 😆

File: submodules/strategyHolidays/strategyHolidays.js [hunk 7]
```diff
@@ -456,6 +576,56 @@ define(function(require) {
...
+					return _.toLower(key);
+				});
+
+			return _.map(csvData, function(row) {
+				var holidayRule = {
+						modified: true,
+						holidayType: row.type
+					},
+					holidayData = {
+						id: 'new-' + Date.now(),
```
Comment (azefiel) [inline submodules/strategyHolidays/strategyHolidays.js:598]: Not sure how holidays are created later on but, wouldn't it make sense to consider objects without an `id` to be new vs generating dummy ids?

Comment (pcandia) [reply submodules/strategyHolidays/strategyHolidays.js:598]: The reason is for editing, you can still edit the holidays after importing it.

Comment (azefiel) [reply submodules/strategyHolidays/strategyHolidays.js:598]: Not sure why an `id` is required though: if an imported holiday is edited, it does not change the fact that it will have to be created on save (which can be determined by the absence of `id`)?

Comment (pcandia) [reply submodules/strategyHolidays/strategyHolidays.js:598]: Hmmm.. once we import it, we store it in local storage along with the existing one. how do you suggest getting the correct "holiday" from the json stored local storage? I originally didn't have the ID, but since the name doesn't have to be unique when importing it, I couldn't find other way to do it.

File: submodules/strategyHolidays/strategyHolidays.js [hunk 8]
```diff
@@ -243,6 +272,97 @@ define(function(require) {
...
+					validateDate = function(date) {
+						var dateArray = _.split(date, ' '),
+							day = _.toInteger(dateArray[1]),
+							isMonthValid = _.includes(months, dateArray[0]),
+							isDayValid = day > 0 && day < 32;
+
+						return _.every([
+							isMonthValid,
+							isDayValid
+						]);
```
[the remark this answers is not in the document]
Comment (pcandia) [reply submodules/strategyHolidays/strategyHolidays.js:309]: I was having issues with advanced, specially as it doesn't support second/third/fourth. I saw ways to add 7 days to the date so I will look into it again and see if i can take advanced as "first" and add the 7 days based on selection.  I will get check out sugarjs again.

File: submodules/strategyHolidays/strategyHolidays.js [hunk 9]
```diff
@@ -236,13 +267,105 @@ define(function(require) {
...
+						validator: function(row) {
+							var start_date = row.start_date,
+								end_date = row.end_date,
+								isTypeValid = _.includes(dateTypes, row.type),
+								isStartDateValid = row.type === 'advanced'
+									? validateAdvancedDate(row.start_date)
+									: validateDate(start_date),
+								isEndDateValid = row.type === 'range'
+									? validateDate(end_date)
+									: true;
```
Comment (azefiel) [inline submodules/strategyHolidays/strategyHolidays.js:353]: This behavior could be extracted into a data structure:
```suggestion
var type = row.type,
	isTypeValid = _.includes(dateTypes, type),
	datesValidator = _.get({
		advanced: {
			start: validateAdvancedDate,
			end: _.stubTrue
		},
		range: {
			start: validateDate,
			end: validateDate
		}
	}, type, {
		start: validateDate,
		end: _.stubTrue
	}),
	isStartDateValid = datesValidator.start(start_date),
	isEndDateValid = datesValidator.end(end_date);
```

Not sure what the other `type` so I've added a default validator but you get the idea.

File: submodules/strategyHolidays/strategyHolidays.js [hunk 10]
```diff
@@ -456,6 +570,74 @@ define(function(require) {
...
+			return _.map(csvData, function(row) {
+				var type = row.type,
+					advancedDateDetails = function formattedAdvancedDate(date) {
+						var dateArray = _.split(row.start_date, ' '),
+							getOrdinal = _
+								.chain(dateArray)
+								.reject(function(value) {
+									return !_.includes(ordinals, value);
+								})
+								.value(),
```
Comment (azefiel) [inline submodules/strategyHolidays/strategyHolidays.js:591]: Same here, `lodash#reject` is the only operation in the sequence, making chaining unnecessary.

File: submodules/strategyHolidays/strategyHolidays.js [hunk 11]
```diff
@@ -236,13 +267,105 @@ define(function(require) {
...
+							}),
+							setDateToValidate = _.isEmpty(ordinalInput)
+								? date
+								: _.replace(date, ordinalInput, 'first'),
+							formattedDate = Sugar.Date.create(setDateToValidate),
+							isValid = formattedDate instanceof Date && !isNaN(formattedDate.getTime());
+
+						return _.every([
+							isOrdinalValid,
+							setDateToValidate,
```
[the remark this answers is not in the document]
Comment (pcandia) [reply submodules/strategyHolidays/strategyHolidays.js:322]: Oops, this was a mistake

File: submodules/strategyHolidays/strategyHolidays.js [hunk 12]
```diff
@@ -236,13 +267,105 @@ define(function(require) {
...
+					isNew: true,
 					callback: function(err, data) {
 						self.appFlags.strategyHolidays.allHolidays.push(data);
-						self.strategyHolidaysListingRender(parent, [data]);
+						self.strategyHolidaysListingRender(parent);
 					}
 				});
 			});
 
+			template.on('click', '.import-csv', function(event) {
```
[the remark this answers is not in the document]
Comment (pcandia) [reply submodules/strategyHolidays/strategyHolidays.js:278]: sure, I will do that.

**Review by azefiel — APPROVED** [review]
Looks good, merge away!
